tests/templates/test_subroutines/test_controlled_sequence.py

```diff
@@ -0,0 +1,408 @@
+    def test_qnode_jax(self, shots, use_jit):
+        """Test that the QNode executes and is differentiable with JAX. The shots
+        argument controls whether autodiff or parameter-shift gradients are used."""
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+
+        dev = qml.device("default.qubit", shots=shots, seed=10)
+        diff_method = "backprop" if shots is None else "parameter-shift"
+        qnode = qml.QNode(self.circuit, dev, interface="jax", diff_method=diff_method)
+        if use_jit:
+            qnode = jax.jit(qnode)
+
+        x = jax.numpy.array(self.x)
+        res = qnode(x)
+        assert qml.math.shape(res) == (16,)
+        assert np.allclose(res, self.exp_result, atol=0.005)
+
+        jac_fn = jax.jacobian(qnode)
+        if use_jit:
+            jac_fn = jax.jit(jac_fn)
+
+        jac = jac_fn(x)
+        assert jac.shape == (16,)
+        assert np.allclose(jac, self.exp_jac, atol=0.006)
+
+    @pytest.mark.torch
+    @pytest.mark.parametrize("shots", [None, 10000])
+    def test_qnode_torch(self, shots):
+        """Test that the QNode executes and is differentiable with Torch. The shots
+        argument controls whether autodiff or parameter-shift gradients are used."""
+        import torch
+
+        dev = qml.device("default.qubit", shots=shots, seed=10)
+        diff_method = "backprop" if shots is None else "parameter-shift"
+        qnode = qml.QNode(self.circuit, dev, interface="torch", diff_method=diff_method)
+
+        x = torch.tensor(self.x, requires_grad=True)
+        res = qnode(x)
+        assert qml.math.shape(res) == (16,)
+        assert qml.math.allclose(res, self.exp_result, atol=0.002)
+
+        jac = torch.autograd.functional.jacobian(qnode, x)
+        assert qml.math.shape(jac) == (16,)
+        assert qml.math.allclose(jac, self.exp_jac, atol=0.006)
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize("shots", [None, 10000])
+    @pytest.mark.xfail(reason="tf gradient doesn't seem to be working, returns ()")
+    def test_qnode_tf(self, shots):
+        """Test that the QNode executes and is differentiable with TensorFlow. The shots
+        argument controls whether autodiff or parameter-shift gradients are used."""
+        import tensorflow as tf
+
+        dev = qml.device("default.qubit", shots=shots, seed=10)
+        diff_method = "backprop" if shots is None else "parameter-shift"
+        qnode = qml.QNode(self.circuit, dev, interface="tf", diff_method=diff_method)
+
+        x = tf.Variable(self.x)
+        with tf.GradientTape() as tape:
+            res = qnode(x)
+
+        assert qml.math.shape(res) == (16,)
+        assert qml.math.allclose(res, self.exp_result, atol=0.002)
+
+        jac = tape.gradient(res, x)
+        assert qml.math.shape(jac) == (16,)
+
+    def test_prod_rx_rx_compiled_circuit(self):
+        """Test that a circuit can execute successfully using qml.compile and
+        a CompositeOp as a base"""
```

Comment:
Do we test this for all templates?

Reply:
No, but changes were made in this PR to allow this to be used with `qml.compile`, so it's relevant here